libraries/Microsoft.Bot.Builder.LanguageGeneration/LGResource.cs

```diff
@@ -46,6 +50,69 @@ public LGResource(IList<LGTemplate> templates, IList<LGImport> imports, string i
         /// </value>
         public IList<LGImport> Imports { get; set; }
 
+        /// <summary>
+        /// update an exist template.
+        /// </summary>
+        /// <param name="templateName">origin template name. the only id of a template.</param>
+        /// <param name="parameters">new params.</param>
+        /// <param name="templateBody">new template body.</param>
+        /// <returns>new LG resource.</returns>
+        public LGResource UpdateTemplate(string templateName, List<string> parameters, string templateBody)
+        {
+            var template = Templates.FirstOrDefault(u => u.Name == templateName);
+            if (template == null)
+            {
+                return this;
+            }
+            var templateNameLine = BuildTemplateNameLine(templateName, parameters);
+            var content = $"{templateNameLine}\r\n{templateBody}";
+            var startLine = template.ParseTree.Start.Line - 1;
+            var stopLine = template.ParseTree.Stop.Line - 1;
+
+            var currentContent = ReplaceContent(OriginalContent, startLine, stopLine, content);
+            return LGParser.Parse(currentContent, Id);
+        }
+
+        /// <summary>
+        /// Add a new template and return LG resource.
+        /// </summary>
+        /// <param name="templateName">new template name.</param>
+        /// <param name="parameters">new params.</param>
+        /// <param name="templateBody">new  template body.</param>
+        /// <returns>new lg resource.</returns>
+        public LGResource AddTemplate(string templateName, List<string> parameters, string templateBody)
+        {
+            var template = Templates.FirstOrDefault(u => u.Name == templateName);
+            if (template != null)
+            {
+                throw new Exception($"template {templateName} already exists.");
+            }
+            var templateNameLine = BuildTemplateNameLine(templateName, parameters);
+
+            var currentContent = $"{OriginalContent}\r\n{templateNameLine}\r\n{templateBody}\r\n";
+            return LGParser.Parse(currentContent, Id);
+        }
+
+        /// <summary>
+        /// Delete an exist template.
+        /// </summary>
+        /// <param name="templateName">which template should delete.</param>
+        /// <returns>return the new lg resource.</returns>
+        public LGResource DeleteTemplate(string templateName)
+        {
+            var template = Templates.FirstOrDefault(u => u.Name == templateName);
+            if (template == null)
+            {
+                return this;
+            }
+
+            var startLine = template.ParseTree.Start.Line - 1;
+            var stopLine = template.ParseTree.Stop.Line - 1;
+
+            var currentContent = ReplaceContent(OriginalContent, startLine, stopLine, null);
```

Comment:
rename currentContent to newContent?  because you are returning a new LGResource